crates/typst/src/layout/grid/layout.rs

```diff
@@ -88,6 +96,65 @@ impl<T: FromValue> FromValue for Celled<T> {
     }
 }
 
+impl<T: Fold> Fold for Celled<T> {
+    fn fold(self, outer: Self) -> Self {
+        match (self, outer) {
+            (Self::Value(inner), Self::Value(outer)) => Self::Value(inner.fold(outer)),
+            (self_, _) => self_,
+        }
+    }
+}
+
+impl<T: Resolve> Resolve for Celled<T> {
+    type Output = ResolvedCelled<T>;
+
+    fn resolve(self, styles: StyleChain) -> Self::Output {
+        match self {
+            Self::Value(value) => ResolvedCelled(Celled::Value(value.resolve(styles))),
+            Self::Func(func) => ResolvedCelled(Celled::Func(func)),
+            Self::Array(values) => ResolvedCelled(Celled::Array(
+                values.into_iter().map(|value| value.resolve(styles)).collect(),
+            )),
+        }
+    }
+}
+
+/// The result of resolving a Celled's value according to styles.
+/// Holds resolved values which depend on each grid cell's position.
+/// When it is a closure, however, it is only resolved when the closure is
+/// called.
+#[derive(Default, Clone)]
+pub struct ResolvedCelled<T: Resolve>(Celled<T::Output>);
```

Comment:
I actually wasn't very happy with `ResolvedCelled` either, and you might see that in earlier versions of this code it was much worse haha. However, I came to believe that this form might be the most minimal and simultaneously correct it can get, for two reasons:

1. If we keep a `Celled<T>` with a (hypothetical) `deep_resolve` method which takes `styles`, we'd always have to repeat the resolve operation for `Celled::Value` and `Celled::Array` for each cell, even when that's totally unneeded since the user specified `#table(stroke: red)`, which in the current design is only resolved once. Of course, there's no way around that for `Celled::Func`, but it'd be a bit of a waste, both of performance and memory.
2. So we could in theory just implement `Resolve` for `Celled<T>` and return a `Celled<T::Output>`. While that would work for `Celled::Value` and `Celled::Array` very nicely, it would fail for `Celled::Func`, since the function would have to return an already resolved value / `T::Output` would have to implement `FromValue` (which isn't the case for `Stroke<Abs>` compared to just `Stroke` aka `Stroke<Length>`).

So the way I found to satisfy both requirements - resolve a single value and an array in advance, but resolve the closure later by having it return a `T` and then resolve into `T::Output` - was to create this separate type (and I kinda abuse the fact that `Celled::Func` is type-erased here, hence why I can use `Celled<T::Output>` even though we will be expecting a `T` from the function later).

I'm not sure if there's a better way to design this, but please let me know if you have a better idea, I'd love to have a more unified solution.

Reply:
Okay, I see. So, it's mostly for performance then, got it. Resolving generally feels cheap, but for strokes it might not actually be that cheap.

Reply:
It's probably not that much of a gain for most typical cases, but it felt like it could be quite considerable at scale, so I feel like it's worth it. Let's keep it in for now and see. :)